kale/prepdata/supergraph_construct.py

```diff
@@ -0,0 +1,261 @@
+                idx = edgetype
+                range_list[-1][1] = i
+                range_list.append([i, i])
+
+        range_list[-1][1] = i + 1
+
+        self.range_list = torch.tensor(range_list)
+
+    def __repr__(self) -> str:
+        return f"SuperVertex(\n    name={self.name}, \n    node_feat={self.node_feat.shape}, \n    edge_index={self.edge_index.shape}, \n    num_edge_type={self.num_edge_type})"
+
+    def add_in_supervertex(self, vertex_name: str):
+        self.in_supervertex_list.append(vertex_name)
+
+    def add_out_supervertex(self, vertex_name: str):
+        self.out_supervertex_list.append(vertex_name)
+
+
+class SuperEdge(object):
+    r"""
+    A superedge structure in GripNet. Each superedge is a bipartite subgraph containing nodes from two categories
+    forming two node sets, connected by edges between them. The superedge can be regards as a heterogeneous graph
+    connecting two supervertices.
+
+    Args:
+        source_supervertex (str): the name of the source supervertex.
+        target_supervertex (str): the name of the target supervertex.
+        edge_index (torch.Tensor): edge indices in COO format with shape [2, #edges]. The first row is the index of
+            source nodes, and the second row is the index of target nodes.
+        edge_weight (torch.Tensor, optional): one-dimensional weight for each edge. Defaults to None.
+    """
+
+    def __init__(
+        self,
+        source_supervertex: str,
+        target_supervertex: str,
+        edge_index: torch.Tensor,
+        edge_weight: torch.Tensor = None,
+    ) -> None:
+
+        self.direction = (source_supervertex, target_supervertex)
+        self.source_supervertex = source_supervertex
+        self.target_supervertex = target_supervertex
+        self.edge_index = edge_index
+        self.edge_weight = edge_weight
+
+    def __repr__(self) -> str:
+        return f"SuperEdges(\n    edge_direction={self.source_supervertex}->{self.target_supervertex}, \n    edge_index={self.edge_index.shape})"
+
+
+class SuperVertexParaSetting(object):
+    r"""Parameter settings for each supervertex.
+
+        Args:
+            supervertex_name (str): the name of the supervertex.
+            internal_feat_channels (int): the dimension of the output of the internal feature layer.
+            internal_agg_channels_list (List[int]): the output dimensions of a sequence of internal aggregation layers.
+            external_agg_channels_dict (Dict[str, int], optional): the dimension of received message vector from parent supervertices.
+                Defaults to None.
+            mode (str, optinal): the allowed gripnet mode--'cat' or 'add'. Defaults to None.
+            num_bases (int, optional): the umber of bases used for basis-decomposition if the
+                supervertex is multi-relational. Defaults to 32.
+            concat_output (bool, optional): whether concatenate the output of each layers. Defaults to True.
```

Comment:
add 'to' between whether and a verb (for all such usage): whether to concatenate ...